### DISCUSSION TITLE
Best way to feed in timestamps?

### DISCUSSION THREAD

**Discussion by IcedQuinn** [thought-machine/please#2218] [Q&A] [answered]
What is the best way to feed in data like build timestamps? For example if a rule is importing from the latest tarball output by upstream simply depending on that will download the file one time but will not re-download it when upstream changes.

What I've experimented with is either having a touch file that holds the stamps and is depended on to force the rule to restart (and then write that from a task runner.) Alternatively it looks like forcing some token rule to rebuild (which say, takes the current time and writes that to a touch file inside as a rule) will force the immediately dependent rules to re-run (and may check their hashes to find that re-running was not needed and everything down the line will not run--this is cool.) But these both feel like very dirty hacks.

Granted depending on `:lastest` is something I'm not a fan of in general. The *cleanest* is to have a task runner update a manifest which please.build then imports via rules and doles out jobs from that. But it changes you from running `please build` to `lets build` (or invoke or fish or etc.)

**Comment by Tatskaari** [comment, marked as answer]
I don't know if this helps, but if you just want to rebuild a target to force Please to download it again, you can use the `--rebuild` flag. Perhaps you can just rebuild all the affected targets in your CI pipeline? I imagine locally it might be easy enough to do something similar when you notice it's out of date. 

Unfortunately, there's no mechanism by which Please could tell if the remote file was changing if the URL doesn't include some sort of versioning information. We'd just have to re-download the file every time you depend on that target. That's not really something I'd want to have Please do. 

Really for this to work nicely you need to have some sort of versioned URL. Having a task runner update a manifest file just pushes the problem on down stream, doesn't it?

**Reply by IcedQuinn** [reply]
When I last faced the problem I did force a rebuild on some token dependencies. It didn't cause a cascade of downstream things to be built though. You had to forcibly expire the token dependencies and then trigger a full build after that. I don't remember specifically what that was about (i think it was build timestamps or the repo hash or something.)

You can get a couple git-specific details but those are magic functions as you can't run workspace commands when preparing rules (Bazel does allow that to a limited extent iirc.) You can harvest all of that and then depend on the resulting manfiest but.. yeah, that introduces a task runner. I'm currently testing out [https://lets-cli.org/](Lets) though i previously used Plan 9 mk for that (the syntax is very concise.)
